modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb

```diff
@@ -61,7 +61,11 @@ def legacy_classes_format(response)
 
   def get_classes(environment)
     future = async_get_classes(environment)
-    future.value!(MAX_CLIENT_TIMEOUT)
+    cache_used = @m.synchronize { !!@etag_cache[environment] } #etags are only available when classes cache is enabled
+    client_timeout = cache_used ? DEFAULT_CLIENT_TIMEOUT : 2 * DEFAULT_CLIENT_TIMEOUT
```

Comment:
With the class cache enabled you need to either restart/reload the server on every code change or call an API, so it changes how environment deployments need to be made. It's also not enabled by default, so I don't think the default behaviour of the smart proxy can be too aggressive in assuming it is.